core/src/test/java/io/parsingdata/metal/token/TieTest.java

```diff
@@ -0,0 +1,278 @@
+        assertTrue(result.succeeded);
+        assertEquals(data.length, result.environment.offset);
+        return result;
+    }
+
+    @Test
+    public void increasing() throws IOException {
+        checkFullParse(INC_PREV_MOD_100, generateIncreasing(1024));
+    }
+
+    private static byte[] generateIncreasing(final int size) {
+        final byte[] data = new byte[size];
+        for (int i = 0; i < data.length; i++) {
+            data[i] = (byte)((i+1) % 100);
+        }
+        return data;
+    }
+
+    @Test
+    public void multiLevelContainers() throws IOException {
+        final byte[] l3Data = generateIncreasing(880);
+        final Token l3Token = INC_PREV_MOD_100;
+        checkFullParse(l3Token, l3Data);
+
+        final byte[] l2Data = flipBlocks(l3Data, 40);
+        final Token l2Token =
+            seq(rep(seq(def("left", con(40)), def("right", con(40)))),
+                tie(l3Token, fold(cat(ref("right"), ref("left")), CAT_REDUCER)));
+        checkFullParse(l2Token, l2Data);
+
+        final byte[] l1Data = prefixSize(deflate(l2Data));
+        final Token l1Token =
+            seq(def("size", con(4)),
+                def("data", last(ref("size"))),
+                tie(l2Token, inflate(last(ref("data")))));
+        final ParseResult result = checkFullParse(l1Token, l1Data);
+        assertEquals(80, result.environment.order.head.asGraph().head.asGraph().head.asGraph().head.asGraph().head.asGraph().head.asValue().asNumeric().intValue());
+    }
+
+    private byte[] flipBlocks(byte[] input, int blockSize) {
+        if ((input.length % (blockSize * 2)) != 0) { throw new RuntimeException("Not supported."); }
+        final byte[] output = input.clone();
+        for (int i = 0; i < output.length; i+= blockSize * 2) {
+            for (int j= 0; j < blockSize; j++) {
+                output[j+i] = input[j+i+blockSize];
+                output[j+i+blockSize] = input[j+i];
+            }
+        }
+        return output;
+    }
+
+    private byte[] deflate(byte[] data) {
+        final Deflater deflater = new Deflater(Deflater.DEFAULT_COMPRESSION, true);
+        deflater.setInput(data);
+        deflater.finish();
+        final byte[] buffer = new byte[data.length * 2];
+        int length = deflater.deflate(buffer);
+        deflater.end();
+        final byte[] output = new byte[length];
+        System.arraycopy(buffer, 0, output, 0, length);
+        return output;
+    }
+
+    private byte[] prefixSize(byte[] data) {
+        final byte[] output = new byte[data.length + 4];
+        output[0] = (byte)((data.length & 0xff000000) >> 24);
+        output[1] = (byte)((data.length & 0xff0000) >> 16);
+        output[2] = (byte)((data.length & 0xff00) >> 8);
+        output[3] = (byte) (data.length & 0xff);
+        System.arraycopy(data, 0, output, 4, data.length);
+        return output;
+    }
+
+    public static ValueExpression inflate(final ValueExpression target) {
```

Comment:
I've moved it to `Util` in metal-core instead, so that it can be used in both core/tests and formats/main. Cleanest solution I could come up with :)